bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/utils/mirroringmetrics/MirroringSpanFactory.java

```diff
@@ -0,0 +1,285 @@
+        try (Scope scope = spanAsScope(span)) {
+          callback.onFailure(throwable);
+        }
+      }
+    };
+  }
+
+  public <T> Callable<T> wrapWithCurrentSpan(final Callable<T> callable) {
+    final Span span = getCurrentSpan();
+    return new Callable<T>() {
+      @Override
+      public T call() throws Exception {
+        try (Scope scope = spanAsScope(span)) {
+          return callable.call();
+        }
+      }
+    };
+  }
+
+  public void asyncCloseSpanWhenCompleted(ListenableFuture<Void> onLastReferenceClosed) {
+    final Span span = MirroringSpanFactory.this.asyncCloseSpan();
+    onLastReferenceClosed.addListener(
+        new Runnable() {
+          @Override
+          public void run() {
+            span.end();
+          }
+        },
+        MoreExecutors.directExecutor());
+  }
+
+  private Span asyncCloseSpan() {
+    return tracer.spanBuilder("asyncClose").startSpan();
+  }
+
+  public <T> T wrapPrimaryOperation(
+      CallableThrowingIOException<T> operationRunner, HBaseOperation operationName)
+      throws IOException {
+    try {
+      return wrapPrimaryOperationAndMeasure(operationRunner, operationName);
+    } catch (InterruptedException e) {
+      assert false;
+      throw new IllegalStateException();
+    }
+  }
+
+  public <T> void wrapPrimaryOperation(
+      CallableThrowingIOAndInterruptedException<T> operationRunner, HBaseOperation operationName)
+      throws IOException, InterruptedException {
+    wrapPrimaryOperationAndMeasure(operationRunner, operationName);
+  }
+
+  public <T> T wrapSecondaryOperation(
+      CallableThrowingIOException<T> operationRunner, HBaseOperation operationName)
+      throws IOException {
+    try {
+      return wrapSecondaryOperationAndMeasure(operationRunner, operationName);
+    } catch (InterruptedException e) {
+      assert false;
+      throw new IllegalStateException();
+    }
+  }
+
+  public <T> T wrapSecondaryOperation(
+      CallableThrowingIOAndInterruptedException<T> operationRunner, HBaseOperation operationName)
+      throws IOException, InterruptedException {
+    return wrapSecondaryOperationAndMeasure(operationRunner, operationName);
+  }
+
+  public <T> FutureCallback<T> wrapReadVerificationCallback(final FutureCallback<T> callback) {
+    return new FutureCallback<T>() {
+      @Override
+      public void onSuccess(@NullableDecl T t) {
+        try (Scope scope = MirroringSpanFactory.this.verificationScope()) {
+          callback.onSuccess(t);
+        }
+      }
+
+      @Override
+      public void onFailure(Throwable throwable) {
+        try (Scope scope = MirroringSpanFactory.this.verificationScope()) {
+          callback.onFailure(throwable);
+        }
+      }
+    };
+  }
+
+  public <T> WriteOperationFutureCallback<T> wrapWriteOperationCallback(
+      final WriteOperationFutureCallback<T> callback) {
+    return new WriteOperationFutureCallback<T>() {
+      @Override
+      public void onFailure(Throwable throwable) {
```

Comment:
why is there no onSuccess here? Please add a  comment or 2.

Reply:
Done.